README.md
plugin.video.last_played
------------------------
Lists what was played most recently, allowing to resume watching directly from the addon.

Before if you wanted to finish viewing something you stopped, you had to remember it and then search, or go through a number of menus to get back to where you were.
Now just open this addon and it will display a list with the last things you watched. Click and it will resume playing directly from the addon.

Usage
-----
When you start playing anything in KODI it will be added to the list of last played.

To remove from the played list something you watched:
Open the list of last played, select the line to remove, and choose "Remove from list" form the context menu. 

Settings:
---------
Number of lines: How many lines to display on screen<br />
Single List: Shows everything on a single list, no matter what or where is was played<br />
Group by Type: Display on separate groups Movies, Shows, Videos, etc<br />
Group by Source: Display separate groups for each addon used to play<br />
Number of lines on top: If grouping is active, display before the groups some of the last played movies or shows<br />
Show Date: Display for each item the date when it was played<br />
Show Time: Display time of play<br />
Play History at Star Movies: The play history can be viewed privately on the www.starmovies.org site<br />
Custom Path: Location of non-library play history. Point all devices to the same location to synch non-library history<br />
Enable Debug: Display an extra menu option allowing to clear the full list, and logs information to help solve eventual errors<br />

Installation
------------
- Download the 5 star repo from https://www.starmovies.org/kodi.zip<br />
- Open Kodi<br />
- Go to `System -> Settings -> Add-ons<br />
- Select Install from zip file` -> Select the downloaded zip file<br />
- Then select -> Install from repository -> Last played addon and Install<br />
- Enjoy :)<br />
 
Release history
---------------
6.0.3 - Fix error clearing history<br />
6.0.1 - Fixes<br />
6.0.0 - Kodi 19<br />
1.3.1 - Fixes<br />
1.3.0 - Setting to allow music<br />
1.2.2 - Update for Kodi 18<br />
1.1.8 - Wildcard Blacklist<br />
1.1.8 - Blacklist and new site<br />
1.0.9 - Skin handling<br />
1.0.8 - New replay method and some minor improvements<br />
1.0.7 - New list format and add 5star-movies<br />
1.0.6 - Configuration options<br />
1.0.5 - Add debug options<br />
1.0.4 - New options: Split by addon and display date and time<br /> 
1.0.3 - Improve metadata info. Aloow item removal. Make it kodi repo compliant<br />
1.0.2 - Set media type on list items<br />
1.0.1 - Initial Release<br />


changelog.txt
[B]6.1.0[/B] - Kodi 20
[B]6.0.3[/B] - Fix error clearing history
[B]6.0.0[/B] - Kodi 19
[B]1.3.1[/B] - Fixes
[B]1.3.0[/B] - Setting to allow music
[B]1.2.2[/B] - Kodi 18 Update
[B]1.2.0[/B] - Monitor videos files only
[B]1.1.10[/B] - New site starmovies.org
[B]1.1.9[/B] - Wildcard Blacklist
[B]1.1.8[/B] - Blacklist and New Site
[B]1.0.8[/B] - New replay method and some minor improvements
[B]1.0.7[/B] - New list format and add 5star-movies
[B]1.0.6[/B] - Configuration options
[B]1.0.5[/B] - Add debug options 
[B]1.0.4[/B] - New options: Split by addon and display date and time 
[B]1.0.3[/B] - Improve metadata info. Aloow item removal. Make it kodi repo compliant
[B]1.0.2[/B] - Set media type on list items
[B]1.0.1[/B] - Initial Release


addon.py
# -*- coding: utf-8 -*-
import sys
import json
import xbmc
import xbmcplugin
import xbmcaddon
import xbmcvfs
import xbmcgui
from xbmcplugin import addDirectoryItem, endOfDirectory
try:
    from urllib.parse import parse_qs, urlencode
except ImportError:
    from urlparse import parse_qs
    from urllib import urlencode

addon = xbmcaddon.Addon()
addon_handle = int(sys.argv[1])
args = parse_qs(sys.argv[2][1:])
menu = args.get('menu', None)
try: list_size = int(addon.getSetting('list_size'))
except Exception: list_size=0
try: top_size = int(addon.getSetting('top_size'))
except Exception: top_size=0
single_list = addon.getSetting('single_list')
group_by = addon.getSetting('group_by')
show_date = addon.getSetting('show_date')
show_time = addon.getSetting('show_time')
enable_debug = addon.getSetting('enable_debug')
lang = addon.getLocalizedString
if addon.getSetting('custom_path_enable') == "true" and addon.getSetting('custom_path') != "":
    txtpath = addon.getSetting('custom_path')
else:
    txtpath = xbmcvfs.translatePath(addon.getAddonInfo('profile'))
txtfile = txtpath + "lastPlayed.json"
imgPath=xbmcvfs.translatePath(xbmcaddon.Addon().getAddonInfo('path'))
group_by_type=lang(30018)

def url(pQuery):
    return sys.argv[0] + '?' + urlencode(pQuery)

def list_items(selGroup, nbrLines):
    xbmcplugin.setContent(addon_handle, "files")
    if xbmcvfs.exists(txtfile):
        f = xbmcvfs.File(txtfile)
        nbr=0 # nbr of line on screen (selected)
        idx=0 # idx of line on the json file
        try: lines = json.load(f)
        except: lines = []
        for line in lines:
            if nbr>=nbrLines: break
            if group_by == group_by_type: group = line["type"]
            else: group = line["source"]
            if len(line)>3 and (group==selGroup or selGroup=='*'):
                nbr=nbr+1
                desc=''
                if show_date == "true": desc = desc + line["date"].strip() + ' '
                if show_time == "true": desc = desc + line["time"].strip() + ' '
                show = ''
                if 'show' in line: show=line["show"] + " "
                if 'season' in line and line["season"]!='' and str(line["season"])!="-1":
                    if 'episode' in line and line["episode"]!='' and str(line["episode"])!="-1":
                        show = show + str(line["season"])+"x"+str(line["episode"]) + " "
                desc=desc + show + line["title"]
                if 'artist' in line and line["artist"]!='': desc=desc+" - "+str(line["artist"])
                xpath=""
                infolabels={'title': desc, 'year': line['year'], "mediatype": line['type'], 'Top250': line['id']}
                li = xbmcgui.ListItem(label=desc)
                li.setInfo('video', infolabels)
                li.setArt({ "poster" : line["thumbnail"].strip() })
                li.setArt({ "thumbnail" : line["thumbnail"].strip() })
                li.setArt({ "fanart" : line["fanart"].strip() })
                li.setProperty('IsPlayable', 'true')
                command = []
                command.append((lang(30008), "RunPlugin(plugin://plugin.video.last_played?menu=remove&id="+str(idx)+")"))
                if line["file"][:6]=="plugin":
                    command.append((lang(30031)+line["source"], "PlayMedia(" + line["file"] + ")"))
                li.addContextMenuItems(command)
                xurl=line["file"]
                if "video" in line and line["video"]!="": xurl=line["video"]
                addDirectoryItem(addon_handle, xurl, li)
            idx = idx + 1
        f.close()
        if single_list == "true" and nbr == 0:
            li = xbmcgui.ListItem(lang(30030))
            li.setProperty('IsPlayable', 'false')
            addDirectoryItem(addon_handle, "", li, isFolder = True)
            
def list_groups():
    if xbmcvfs.exists(txtfile):
        groups = []
        f = xbmcvfs.File(txtfile)
        try:
            lines = json.load(f);
        except Exception:
            lines = []
        for line in lines:
            if len(line)>5:
                if group_by == group_by_type:
                    group = line["type"]
                else:
                    group = line["source"]
                if group not in groups:
                    groups.append(group)
                    if group_by == group_by_type:
                        nm = group
                        ic = imgPath+'/resources/' + group + '.png'
                    else:
                        nm = group
                        ads = group.split("/")
                        if len(ads) > 2: nm = ads[2]
                        if group==lang(30002): ic = imgPath+'/resources/movie.png'
                        elif group==lang(30003): ic = imgPath+'/resources/episode.png'
                        elif group==lang(30004): ic = imgPath+'/resources/musicvideo.png'
                        else:
                            try:
                                la = xbmcaddon.Addon(nm)
                                nm = la.getAddonInfo('name')
                                ic = la.getAddonInfo('icon')
                            except Exception:
                                ic = imgPath+'/resources/addons.png'
                    li = xbmcgui.ListItem(nm)
                    li.setArt({'thumbnail':ic})
                    addDirectoryItem(addon_handle, url({'menu': group}), li, True)
        f.close()

# Menu
if menu is None or menu[0]=="top":
    if single_list == "true":
        list_items("*", list_size)
    else:
        xbmcplugin.setContent(addon_handle, "menu")
        list_items("*", top_size)
        if top_size>0: addDirectoryItem(addon_handle, "", xbmcgui.ListItem(label=""), False)

        list_groups()
    if enable_debug	== "true":
        addDirectoryItem(addon_handle, url({'menu': 'showlist'}), xbmcgui.ListItem(lang(30014)), True)
        if xbmcvfs.exists(txtfile):
            addDirectoryItem(addon_handle, url({'menu': 'deletelist'}), xbmcgui.ListItem(lang(30015)), True)
        else:
            addDirectoryItem(addon_handle, url({'menu': 'deletelist'}), xbmcgui.ListItem(lang(30016)), True)
    endOfDirectory(addon_handle)
elif menu[0] == 'remove':
    lid = args.get('id', None)
    if xbmcvfs.exists(txtfile) and lid is not None:
        f = xbmcvfs.File(txtfile)
        lines = json.load(f)
        f.close()
        osz = len(lines)
        lines.remove(lines[int(lid[0])])
        # to avoid accidental cleaning, update empty file only it had only one line before
        if len(lines)>0 or osz==1:
            f = xbmcvfs.File(txtfile, 'w')
            json.dump(lines, f)
            f.close()
        xbmc.executebuiltin("ActivateWindow(Videos,plugin://plugin.video.last_played?menu=top)")
elif menu[0] == 'showlist':
    if xbmcvfs.exists(txtfile):
        f = xbmcvfs.File(txtfile)
        lines = json.load(f)
        f.close()
        for line in lines:
            addDirectoryItem(addon_handle, url({}), ListItem(str(line)), False)
        endOfDirectory(addon_handle)
elif menu[0] == 'deletelist':
    if xbmcvfs.exists(txtfile):
        lines = []
        f = xbmcvfs.File(txtfile, 'w')
        json.dump(lines, f)
        f.close()
else:
    list_items(menu[0], list_size)
    endOfDirectory(addon_handle)


default.py
# -*- coding: utf-8 -*-
import os, time
import json
import ssl
import xbmc
import xbmcaddon
import xbmcvfs

try:
    from urllib.parse import quote, unquote
    import urllib.request
    python="3"
except ImportError:
    from urllib import quote, unquote
    import urllib2
    python="2"

addon = xbmcaddon.Addon()
enable_debug = addon.getSetting('enable_debug')
if addon.getSetting('custom_path_enable') == "true" and addon.getSetting('custom_path') != "":
    txtpath = addon.getSetting('custom_path')
else:
    txtpath = xbmcvfs.translatePath(addon.getAddonInfo('profile'))
    if not os.path.exists(txtpath):
        os.makedirs(txtpath)
txtfile = txtpath + "lastPlayed.json"
starmovies = addon.getSetting('starmovies')
lang = addon.getLocalizedString

class LP:
    title = ""
    year = ""
    thumbnail = ""
    fanart = ""
    showtitle = ""
    season = ""
    episode = ""
    DBID = ""
    type = ""
    file=""
    video=""
    artist=""
    vidPos=1
    vidTot=1000

lp=LP()

player_monitor = xbmc.Monitor()

def getRequest2(url):
    try:
        context = ssl._create_unverified_context()
        request = urllib2.Request(url)
        response = urllib2.urlopen(request, context=context)
        return response
    except:
        pass

def getRequest3(url):
    try:
        req = urllib.request.Request(url)
        with urllib.request.urlopen(req) as response:
            return(response.read())
    except:
        pass
        
# Builds JSON request with provided json data
def buildRequest(method, params, jsonrpc='2.0', rid='1'):
    request = { 'jsonrpc' : jsonrpc, 'method' : method, 'params' : params, 'id' : rid }
    return request

# Checks JSON response and returns boolean result
def checkReponse(response):
    result = False
    if ( ('result' in response) and ('error' not in response) ):
        result = True
    return result

# Executes JSON request and returns the JSON response
def JSexecute(request):
    request_string = json.dumps(request)
    response = xbmc.executeJSONRPC(request_string)
    if ( response ):
        response = json.loads(response)
    return response

# Performs single JSON query and returns result boolean, data dictionary and error string
def JSquery(request):
    result = False
    data = {}
    error = ''
    if ( request ):
        response = JSexecute(request)
        if ( response ):
            result = checkReponse(response)
            if ( result ):
                data = response['result']
            else: error = response['error']
    return (result, data, error)

def send2starmovies(line):
    if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (starmovies) "+str(line), 3)
    if LP.vidPos/LP.vidTot<0.8: return
    wid = 0
    if line["id"]!="": wid = int(line["id"])
    if line["type"]=="movie": typ="M"
    elif line["type"]=="episode": typ="S"
    elif line["type"]=="song": typ="P"
    else: typ="V"
    if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (starmovies) "+str(addon.getSetting('smovies')), 3)

    if typ=="M" and addon.getSetting('smovies') != "true": return
    if typ=="S" and addon.getSetting('stv') != "true": return
    if typ=="V" : return
    if typ=="P" : return

    imdbId = ""
    tvdbId = ""
    orgTitle = ""
    showTitle = line["show"]
    season = line["season"]
    episode = line["episode"]
    thumbnail = line["thumbnail"]
    fanart = line["fanart"]

    if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (starmovies) "+str(wid), 3)
    if wid>0:
        if typ=="M":
            request = buildRequest('VideoLibrary.GetMovieDetails', {'movieid' : wid, 'properties' : ['imdbnumber', 'originaltitle']})
            result, data = JSquery(request)[:2]
            if ( result and 'moviedetails' in data ):
                imdbId = data['moviedetails']["imdbnumber"]
                orgTitle = data['moviedetails']["originaltitle"]
        elif typ=="S":
            request = buildRequest('VideoLibrary.GetEpisodeDetails', {'episodeid' : wid, 'properties' : ['tvshowid', 'season', 'episode']})
            result, data = JSquery(request)[:2]
            if ( result and 'episodedetails' in data ):
                season = data['episodedetails']["season"]
                episode = data['episodedetails']["episode"]
                request = buildRequest('VideoLibrary.GetTvShowDetails', {'tvshowid' : data['episodedetails']["tvshowid"], 'properties' : ['imdbnumber', 'originaltitle']})
                result, data = JSquery(request)[:2]
                if ( result and 'tvshowdetails' in data ):
                    showTitle = data['tvshowdetails']["label"]
                    orgTitle = data['tvshowdetails']["originaltitle"]
                    tvdbId = data['tvshowdetails']["imdbnumber"]

    xvideo = line["file"]
    if "video" in line and line["video"]!="": xvideo = line["video"]
    url = "https://www.starmovies.org/WebService.asmx/kodiWatch?tmdbId="
    url = url + "&tvdbId=" + tvdbId
    url = url + "&imdbId=" + imdbId
    url = url + "&kodiId=" + str(wid)
    url = url + "&title=" + quote(line["title"].encode("utf-8"))
    url = url + "&orgtitle=" + quote(orgTitle.encode("utf-8"))
    url = url + "&year=" + str(line["year"])
    url = url + "&source=" + quote(line["source"].encode("utf-8"))
    url = url + "&type=" + typ
    url = url + "&usr=" + quote(addon.getSetting('TMDBusr').encode("utf-8"))
    url = url + "&pwd=" + addon.getSetting('TMDBpwd')
    url = url + "&link=" + quote(xvideo.encode("utf-8"))
    url = url + "&thumbnail=" + quote(line["thumbnail"].encode("utf-8"))
    url = url + "&fanart=" + quote(line["fanart"].encode("utf-8"))
    url = url + "&showtitle=" + quote(showTitle.encode("utf-8"))
    url = url + "&season=" + str(season)
    url = url + "&episode=" + str(episode)
    url = url + "&version=1.22"
    url = url + "&date=" + line["date"]
    if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (starmovies) "+url, 3)
    if python=="3":
        response = getRequest3(url)
    else:
        response = getRequest2(url)
    if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (starmovies) response:"+str(response), 3)

def videoEnd():
    retry=1
    xsource=''
    while xsource=='' and retry<50:
        xsource = xbmc.getInfoLabel('ListItem.Path')
        retry=retry+1
        time.sleep(0.1)

    if xsource=='': xsource="player"
    xtitle = lp.title
    xyear = lp.year
    xartist = lp.artist
    xid = lp.DBID
    xtype = lp.type
    xfanart = unquote(lp.fanart).replace("image://","").rstrip("/")
    xthumb = unquote(lp.thumbnail).replace("image://","").rstrip("/")	
    if ".jpg" not in xthumb.lower(): xthumb=xfanart
    xfile = lp.file.strip()
    xvideo = lp.video.strip()

    try:
        xshow = lp.showtitle
        xseason = lp.season
        xepisode = lp.episode
    except:
        pass

    if xid!="" and int(xid)>0:
        if xtype=="movie": xsource=lang(30002)
        elif xtype=="episode": xsource=lang(30003)
        elif xtype=="musicvideo": xsource=lang(30004)
        else: xsource=xtype
    else:
        ads = xsource.split("/")
        if len(ads) > 2: xsource = ads[2]

    # if source is on blacklist, do not keep
    if xtype=="movie" and addon.getSetting('movies') != "true": return
    if xtype=="episode" and addon.getSetting('tv') != "true": return
    if xtype=="song" and addon.getSetting('music') != "true": return
    if xtype!="movie" and xtype!="episode" and xtype!="song" and addon.getSetting('videos') != "true": return

    # if source is on blacklist, do not keep
    if addon.getSetting('blackadddon').lower().find(xsource.lower())>=0: return
    if addon.getSetting('blackfolder')!="":
        for dir in addon.getSetting('blackfolder').lower().split(","):
            if xsource.lower().find(dir)>=0: return
    if addon.getSetting('blackvideo')!="":
        for vid in addon.getSetting('blackvideo').lower().split(","):
            if xtitle.lower().find(vid)>=0: return
    
    if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (end source) "+xsource, 3)
    if xbmcvfs.exists(txtfile):
        f = xbmcvfs.File(txtfile)
        try: lines = json.load(f)
        except: lines = []
        f.close()
    else: lines = []

    replay = "N"
    for line in lines:
        if xfile!="" and xfile==line["file"]: replay = "S"
        if replay == "S":
            lines.remove(line)
            line.update({"date": time.strftime("%Y-%m-%d")})
            line.update({"time": time.strftime("%H:%M:%S")})
            lines.insert(0, line)
            replay = "S"
            if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (end final replay) "+str(line), 3)
            if starmovies	== "true": send2starmovies(line)
            break

    if replay=="N":
        newline = {"source":xsource, "title":xtitle, "year":xyear, "artist":xartist, "file":xfile, "video":xvideo, "id":xid, "type":xtype,"thumbnail":xthumb, "fanart":xfanart, "show":xshow, "season":xseason, "episode":xepisode, "date":time.strftime("%Y-%m-%d"), "time":time.strftime("%H:%M:%S") }
        lines.insert(0, newline)
        if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (end final play) "+str(newline), 3)
        if starmovies	== "true": send2starmovies(newline)
        if len(lines)>100:
            del lines[len(lines)-1]

    if len(lines)>0:
        f = xbmcvfs.File(txtfile, 'w')
        json.dump(lines, f)
        f.close()
          
class KodiPlayer(xbmc.Player):
    def __init__(self, *args, **kwargs):
        kplayer=xbmc.Player.__init__(self)

    @classmethod
    def onPlayBackEnded(self):
        videoEnd()

    @classmethod
    def onPlayBackStopped(self):
        videoEnd()

    def onPlayBackStarted(self):
        if xbmc.getCondVisibility('Player.HasMedia'):
            lp.video = self.getPlayingFile()
            request = {"jsonrpc": "2.0", "method": "Player.GetItem", "params": { "properties": ["title", "year", "thumbnail", "fanart", "showtitle", "season", "episode", "file"], "playerid": 1 }, "id": "VideoGetItem"}
            result, data = JSquery(request)[:2]
            if(len(data)==0):
                request = {"jsonrpc": "2.0", "method": "Player.GetItem", "params": { "properties": ["title", "album", "artist", "duration", "thumbnail", "file", "fanart", "streamdetails"], "playerid": 0 }, "id": "AudioGetItem"}
                result, data = JSquery(request)[:2]
            if len(data)>0:
                item=data["item"]
                if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (start play) "+str(item), 3)
                if "title" in item: lp.title = item["title"]
                if lp.title=="" and "label" in item: lp.title = item["label"]		
                if "year" in item: lp.year = item["year"]
                if "thumbnail" in item: lp.thumbnail = item["thumbnail"]
                if "fanart" in item: lp.fanart = item["fanart"]
                if "showtitle" in item: lp.showtitle = item["showtitle"]
                if "season" in item and item["season"]>0: lp.season = item["season"]
                if "episode" in item and item["episode"]>0: lp.episode = item["episode"]
                if "id" in item: lp.DBID = item["id"]
                if "type" in item: lp.type = item["type"]
                if "file" in item: lp.file = item["file"]
                if "artist" in item: lp.artist = item["artist"]

class KodiRunner:
    player = KodiPlayer()

    while not player_monitor.abortRequested():
        if player.isPlaying():
            LP.vidPos=player.getTime()
            LP.vidTot=player.getTotalTime()
        player_monitor.waitForAbort(1)

    del player
